src/python/gudhi/sklearn/rips_persistence.py

```diff
@@ -73,36 +84,90 @@ def fit(self, X, Y=None):
+        # Should we use threshold in the sparse case?
+        num_collapses = self.num_collapses
+        input_type = self.input_type
+        threshold = self.threshold
+        n = inp.shape[0] if input_type == 'distance coo_matrix' else len(inp)
+        max_dimension = min(max(self.dim_list_), max(0, n-3))
+        # Ripser needs to encode simplices and coefficients in 128 bits, which may not always fit
+        # Instead of a 256 bit version which may not always suffice either, fall back to SimplexTree
+        use_simplex_tree = math.comb(n, min(n // 2, max_dimension + 2)) >= (1 << (128 - (self.homology_coeff_field - 2).bit_length()))
+        if num_collapses == 'auto':
+            num_collapses = 1 if max_dimension > (not use_simplex_tree) else 0
+            # or num_collapses=max_dimension-1 maybe?
+        elif max_dimension == 0:
+            num_collapses = 0
+
+        # Points -> distance matrix
+        if input_type == 'point cloud':
+            if threshold < float('inf'):
+                # Hope that the user gave a useful threshold
+                tree = cKDTree(inp)
+
+                ## V1: Returns self-loops and every edge twice (symmetry)
+                # inp = tree.sparse_distance_matrix(tree, max_distance=threshold, output_type="coo_matrix")
+                # mask = inp.row < inp.col
+                # inp = coo_matrix((inp.data[mask], (inp.row[mask], inp.col[mask])), shape=inp.shape)
+
+                # V2: Gets the right edges, but forgets the distances
+                pairs = tree.query_pairs(r=threshold, output_type='ndarray')
+                data = np.ravel(np.linalg.norm(np.diff(inp[pairs], axis=1), axis=-1))
+                inp = coo_matrix((data, (pairs[:,0], pairs[:,1])), shape=(n,)*2)
+
+                input_type = 'distance coo_matrix'
+            else:
+                inp = squareform(pdist(inp))
+                input_type = 'full distance matrix'
+
+        # Dense -> sparse
+        if input_type in ('full distance matrix', 'lower distance matrix'):
+            # After this filtration value, all complexes are cones, nothing happens
+            if input_type == 'full distance matrix':
+                inp = np.asarray(inp)
+                cone_radius = inp.max(-1).min()
+            else:
+                cone_radius = _lower_cone_radius(inp)
+            sparsify = use_simplex_tree or num_collapses > 0 or threshold < cone_radius # heuristic
+            threshold = min(threshold, cone_radius)
+            if sparsify:
+                # For 'full' we could use i, j = np.triu_indices_from(inp, k=1), etc
+                i, j, f = _lower_to_coo(inp, threshold)
+                inp = coo_matrix((f, (i, j)), shape=(n,) * 2)
+                input_type = 'distance coo_matrix'
+
+        if num_collapses > 0:
+            assert input_type == 'distance coo_matrix'
+            inp = reduce_graph(inp, num_collapses)
+
+        if use_simplex_tree:
+            from gudhi import SimplexTree
```

Comment:
```python
from .. import SimplexTree
```
and at the top of the file seems to be more _pythonic_

Reply:
For something like SimplexTree that is already imported at the top level anyway, indeed (I thought I had already moved it, but that was probably another import).